pynamodb startswith works with moto but not under pytest

environment:
windows 10 64 bit python 3.6

see zip for code to reproduce issue.
do a virtualenv and install requirements. model, test_model and test_dynamodb contain the same code, except that the 2nd and 3rd are wrapped in order for pytest to work. The 3rd, doesn't use moto, but a dynamodb local listening on localhost 8000.
If you execute
 python model.py,
 everything works correctly.
if, in the test dir, you execute
pytest
the 'query' call using moto mock which uses the startswith, fails with
============================= test session starts =============================
platform win32 -- Python 3.6.4, pytest-4.0.2, py-1.7.0, pluggy-0.8.0
rootdir: C:\Users\FIM514, inifile: pytest.ini
collected 1 item

test_model.py F                                                          [100%]

================================== FAILURES ===================================
______________________________ MyTest.test_main _______________________________

self = <test.test_model.MyTest testMethod=test_main>

    def test_main(self):
        model.create_table(wait=True)
        model(f1='1', f2='2').save()
        print(next(model.query('1', model.f2=='2')))
>       print(next(model.query('1', model.f2.startswith('2'))))

test_model.py:24:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\envs\temp\lib\site-packages\pynamodb\pagination.py:173: in __next__
    self._get_next_page()
..\envs\temp\lib\site-packages\pynamodb\pagination.py:158: in _get_next_page
    page = next(self.page_iter)
..\envs\temp\lib\site-packages\pynamodb\pagination.py:103: in __next__
    page = self._operation(*self._args, **self._kwargs)
..\envs\temp\lib\site-packages\pynamodb\connection\table.py:257: in query
    select=select)
..\envs\temp\lib\site-packages\pynamodb\connection\base.py:1409: in query
    return self.dispatch(QUERY, operation_kwargs)
..\envs\temp\lib\site-packages\pynamodb\connection\base.py:313: in dispatch
    data = self._make_api_call(operation_name, operation_kwargs)
..\envs\temp\lib\site-packages\pynamodb\connection\base.py:362: in _make_api_call
    proxies=proxies,
..\envs\temp\lib\site-packages\requests\sessions.py:646: in send
    r = adapter.send(request, **kwargs)
..\envs\temp\lib\site-packages\responses.py:622: in unbound_on_send
    return self._on_request(adapter, request, *a, **kwargs)
..\envs\temp\lib\site-packages\responses.py:600: in _on_request
    response = adapter.build_response(request, match.get_response(request))
..\envs\temp\lib\site-packages\moto\core\models.py:151: in get_response
    result = self.callback(request)
..\envs\temp\lib\site-packages\moto\core\utils.py:173: in __call__
    result = self.callback(request, request.url, request.headers)
..\envs\temp\lib\site-packages\moto\core\responses.py:117: in dispatch
    return cls()._dispatch(*args, **kwargs)
..\envs\temp\lib\site-packages\moto\core\responses.py:200: in _dispatch
    return self.call_action()
..\envs\temp\lib\site-packages\moto\core\utils.py:264: in _wrapper
    response = f(*args, **kwargs)
..\envs\temp\lib\site-packages\moto\dynamodb2\responses.py:64: in call_action
    response = getattr(self, endpoint)()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <moto.dynamodb2.responses.DynamoHandler object at 0x000002DC6554C6D8>

    def query(self):
        name = self.body['TableName']
        # {u'KeyConditionExpression': u'#n0 = :v0', u'ExpressionAttributeValues': {u':v0': {u'S': u'johndoe'}}, u'ExpressionAttributeNames': {u'#n0': u'username'}}
        key_condition_expression = self.body.get('KeyConditionExpression')
        projection_expression = self.body.get('ProjectionExpression')
        expression_attribute_names = self.body.get('ExpressionAttributeNames', {})
        filter_expression = self.body.get('FilterExpression')
        expression_attribute_values = self.body.get('ExpressionAttributeValues', {})

        if projection_expression and expression_attribute_names:
            expressions = [x.strip() for x in projection_expression.split(',')]
            for expression in expressions:
                if expression in expression_attribute_names:
                    projection_expression = projection_expression.replace(expression, expression_attribute_names[expression])

        filter_kwargs = {}

        if key_condition_expression:
            value_alias_map = self.body.get('ExpressionAttributeValues', {})

            table = self.dynamodb_backend.get_table(name)

            # If table does not exist
            if table is None:
                return self.error('com.amazonaws.dynamodb.v20120810#ResourceNotFoundException',
                                  'Requested resource not found')

            index_name = self.body.get('IndexName')
            if index_name:
                all_indexes = (table.global_indexes or []) + \
                    (table.indexes or [])
                indexes_by_name = dict((i['IndexName'], i)
                                       for i in all_indexes)
                if index_name not in indexes_by_name:
                    raise ValueError('Invalid index: %s for table: %s. Available indexes are: %s' % (
                        index_name, name, ', '.join(indexes_by_name.keys())
                    ))

                index = indexes_by_name[index_name]['KeySchema']
            else:
                index = table.schema

            reverse_attribute_lookup = dict((v, k) for k, v in
                                            six.iteritems(self.body.get('ExpressionAttributeNames', {})))

            if " AND " in key_condition_expression:
                expressions = key_condition_expression.split(" AND ", 1)

                index_hash_key = [key for key in index if key['KeyType'] == 'HASH'][0]
                hash_key_var = reverse_attribute_lookup.get(index_hash_key['AttributeName'],
                                                            index_hash_key['AttributeName'])
                hash_key_regex = r'(^|[\s(]){0}\b'.format(hash_key_var)
                i, hash_key_expression = next((i, e) for i, e in enumerate(expressions)
                                              if re.search(hash_key_regex, e))
                hash_key_expression = hash_key_expression.strip('()')
                expressions.pop(i)

                # TODO implement more than one range expression and OR operators
                range_key_expression = expressions[0].strip('()')
                range_key_expression_components = range_key_expression.split()
                range_comparison = range_key_expression_components[1]

                if 'AND' in range_key_expression:
                    range_comparison = 'BETWEEN'
                    range_values = [
                        value_alias_map[range_key_expression_components[2]],
                        value_alias_map[range_key_expression_components[4]],
                    ]
                elif 'begins_with' in range_key_expression:
                    range_comparison = 'BEGINS_WITH'
                    range_values = [
>                       value_alias_map[range_key_expression_components[1]],
                    ]
E                   KeyError: '(#1,'

..\envs\temp\lib\site-packages\moto\dynamodb2\responses.py:382: KeyError
---------------------------- Captured stdout call -----------------------------
table<1, 2>


the query call in test_dynamodb, once I start dynamodb-local so I don't use moto mock, works correctly.
I expect the 2nd case to be working correctly too.
[test.zip](https://github.com/spulec/moto/files/2679965/test.zip)

--- a/moto/dynamodb2/responses.py
+++ b/moto/dynamodb2/responses.py
@@ -481,7 +481,7 @@ class DynamoHandler(BaseResponse):
                     ]
                 elif "begins_with" in range_key_expression:
                     range_comparison = "BEGINS_WITH"
-                    range_values = [value_alias_map[range_key_expression_components[1]]]
+                    range_values = [value_alias_map[range_key_expression_components[-1]]]
                 else:
                     range_values = [value_alias_map[range_key_expression_components[2]]]
             else:

--- a/moto/dynamodb2/responses.py
+++ b/moto/dynamodb2/responses.py
@@ -438,9 +438,12 @@ class DynamoHandler(BaseResponse):
                 all_indexes = (table.global_indexes or []) + (table.indexes or [])
                 indexes_by_name = dict((i["IndexName"], i) for i in all_indexes)
                 if index_name not in indexes_by_name:
-                    raise ValueError(
-                        "Invalid index: %s for table: %s. Available indexes are: %s"
-                        % (index_name, name, ", ".join(indexes_by_name.keys()))
+                    er = "com.amazonaws.dynamodb.v20120810#ResourceNotFoundException"
+                    return self.error(
+                        er,
+                        "Invalid index: {} for table: {}. Available indexes are: {}".format(
+                            index_name, name, ", ".join(indexes_by_name.keys())
+                        ),
                     )
 
                 index = indexes_by_name[index_name]["KeySchema"]

--- a/moto/dynamodb2/responses.py
+++ b/moto/dynamodb2/responses.py
@@ -481,7 +481,9 @@ class DynamoHandler(BaseResponse):
                     ]
                 elif "begins_with" in range_key_expression:
                     range_comparison = "BEGINS_WITH"
-                    range_values = [value_alias_map[range_key_expression_components[-1]]]
+                    range_values = [
+                        value_alias_map[range_key_expression_components[-1]]
+                    ]
                 else:
                     range_values = [value_alias_map[range_key_expression_components[2]]]
             else:

--- a/tests/test_dynamodb2/test_dynamodb.py
+++ b/tests/test_dynamodb2/test_dynamodb.py
@@ -2655,14 +2655,15 @@ def test_query_by_non_exists_index():
         ],
     )
 
-    with assert_raises(ValueError) as ex:
+    with assert_raises(ClientError) as ex:
         dynamodb.query(
             TableName="test",
             IndexName="non_exists_index",
             KeyConditionExpression="CarModel=M",
         )
 
-    str(ex.exception).should.equal(
+    ex.exception.response["Error"]["Code"].should.equal("ResourceNotFoundException")
+    ex.exception.response["Error"]["Message"].should.equal(
         "Invalid index: non_exists_index for table: test. Available indexes are: test_gsi"
     )
 

--- a/tests/test_dynamodb2/test_dynamodb.py
+++ b/tests/test_dynamodb2/test_dynamodb.py
@@ -2631,6 +2631,43 @@ def test_scan_by_non_exists_index():
 
 
 @mock_dynamodb2
+def test_query_by_non_exists_index():
+    dynamodb = boto3.client("dynamodb", region_name="us-east-1")
+
+    dynamodb.create_table(
+        TableName="test",
+        KeySchema=[{"AttributeName": "id", "KeyType": "HASH"}],
+        AttributeDefinitions=[
+            {"AttributeName": "id", "AttributeType": "S"},
+            {"AttributeName": "gsi_col", "AttributeType": "S"},
+        ],
+        ProvisionedThroughput={"ReadCapacityUnits": 1, "WriteCapacityUnits": 1},
+        GlobalSecondaryIndexes=[
+            {
+                "IndexName": "test_gsi",
+                "KeySchema": [{"AttributeName": "gsi_col", "KeyType": "HASH"}],
+                "Projection": {"ProjectionType": "ALL"},
+                "ProvisionedThroughput": {
+                    "ReadCapacityUnits": 1,
+                    "WriteCapacityUnits": 1,
+                },
+            }
+        ],
+    )
+
+    with assert_raises(ValueError) as ex:
+        dynamodb.query(
+            TableName="test",
+            IndexName="non_exists_index",
+            KeyConditionExpression="CarModel=M",
+        )
+
+    str(ex.exception).should.equal(
+        "Invalid index: non_exists_index for table: test. Available indexes are: test_gsi"
+    )
+
+
+@mock_dynamodb2
 def test_batch_items_returns_all():
     dynamodb = _create_user_table()
     returned_items = dynamodb.batch_get_item(

